zebra-state/src/service/finalized_state/zebra_db/shielded.rs

```diff
@@ -289,6 +290,32 @@ impl ZebraDb {
         }
     }
 
+    /// Get the sapling note commitment subtress for the finalized tip.
+    #[allow(clippy::unwrap_in_result)]
+    fn sapling_subtree_for_tip(&self) -> Option<NoteCommitmentSubtree<sapling::tree::Node>> {
+        let tip_height = self.finalized_tip_height().unwrap_or(Height(0));
+        let sapling_activation_height = NetworkUpgrade::Sapling
+            .activation_height(self.network())
+            .expect("sapling activation height must exist");
+
+        if tip_height >= sapling_activation_height {
+            let sapling_nct_handle = self.db.cf_handle("sapling_note_commitment_tree").unwrap();
+
+            let tip_note_commitment_tree: Arc<sapling::tree::NoteCommitmentTree> = self
+                .zs_last_key_value(&sapling_nct_handle)
+                .map(|(_key, value): (RawBytes, _)| Arc::new(value))
+                .expect("Sapling note commitment tree must exist if there is a finalized tip");
+
+            if let Some((subtree_index, node)) =
+                tip_note_commitment_tree.completed_subtree_index_and_root()
```

Comment:
This method might not do what you think it does, you might want to check the docs. (@arya2 is the best person to explain the details, because he wrote it.)

Reply:
if the details are not clear from the code/documentation we should make sure we capture those also

Reply:
I think we might need to use `subtree_index()` to get the `subtree_index` but unsure how to get the `node` yet.

Reply:
> This method might not do what you think it does, you might want to check the docs. (@arya2 is the best person to explain the details, because he wrote it.)

`completed_subtree_index_and_root()` was meant for finding subtree roots during a format upgrade or when adding verified blocks to the state, could it get the last subtree on disk instead?

Reply:
> if the details are not clear from the code/documentation we should make sure we capture those also

What the method does is clear to me, but the intended usage is not mentioned in the docs:
https://github.com/ZcashFoundation/zebra/blob/cdddbcc7fe1d114fcfc8ce169bf93fdfd2ac4a67/zebra-chain/src/orchard/tree.rs#L537-L538

It might also help to say that the method returns `None` most of the time. (It only returns `Some` for `1 / 2^16` notes.)

@arya2 is this documentation something you could fix?

Reply:
> could it get the last subtree on disk instead?

Is this what we actually want though?

We don't know what the method is meant to do, because the documentation on `note_commitment_trees()` is unclear and needs to be updated.

But looking at how `note_commitment_trees()` is called, it is only used in this one place, where the subtrees are ignored:
https://github.com/ZcashFoundation/zebra/blob/cdddbcc7fe1d114fcfc8ce169bf93fdfd2ac4a67/zebra-state/src/service/finalized_state.rs#L245-L250

But it looks like it is meant to be equivalent to `prev_note_commitment_trees`, which is populated by `Chain::treestate()` (which also needs documentation). So I had to look all the way back to:
https://github.com/ZcashFoundation/zebra/blob/cdddbcc7fe1d114fcfc8ce169bf93fdfd2ac4a67/zebra-state/src/service/non_finalized_state/chain.rs#L694-L701

So let's assume `note_commitment_trees()` is like `treestate()`, and it should return the trees for the tip block, and the subtrees that end at the tip block. So the subtrees will usually be `None`. (These methods could use the same name to make their matching behaviour clearer.)

If we make the suggested changes to alway get the last subtree, the code will be incorrect, because the last subtree usually ends before the tip block.

Reply:
It could filter the return value unless the end height matches the finalized tip height.

I think documenting that only used for writing and always `None` on read would be good too.